### PR TITLE
new Ensemble and changes

### PR DESCRIPTION
**Pull request by nivfono**

### CODE REVIEW

File: src/model/ensemble_v2.py [hunk 1]
```diff
@@ -74,15 +75,49 @@ def merge_textparts(self, char_textpart: TextPart, word_textpart: TextPart, char
...
-                if char_pred.value == word_pred.value:
-                    score = char_model_weight * char_pred.score + (1 - char_model_weight) * word_pred.score
-                    res_preds.append(Prediction(char_pred.value, score))
-
-        return TextPart('?', list(sorted(res_preds, key=lambda x: x.score, reverse=True)))
+        new_scores=self.calculate_new_scores(char_textpart.predictions, word_textpart.predictions, char_model_weight, 1-char_model_weight)
+        res=[Prediction(value,score) for value,score in new_scores.items()]
+        return TextPart('?', list(sorted(res, key=lambda x: x.score, reverse=True)))
+
+    def calculate_new_scores(self, list1, list2, weight1, weight2):
```
Comment (Eldar-Ka) [inline src/model/ensemble_v2.py:82]: please add docomentation

File: src/model/single_char_model.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import itertools
+from typing import List
+
+from src.classes.model_result import ModelResult
+from src.classes.prediction import Prediction
+from src.classes.text_part import TextPart
+from src.model.standard_model import StandardModel
+
+
+class single_char_model(StandardModel):
```
Comment (Eldar-Ka) [inline src/model/single_char_model.py:10]: why do we need this model?

File: src/space_predictor/Iterative_space_predictor.py [hunk 3]
```diff
@@ -13,19 +15,34 @@ def genText(self,model:Model,text:str,threshold=0.1)->(str,Model):
...
-                        new_model_lst.pop(index-time_of_spaces)
-                        time_of_spaces+=1
+                for prediction in text_part.predictions:
+                    if prediction.value=='':
+                        indeces_of_spaces.append(index_in_text)
+                index_in_text+=1
+        new_text=self.replace_with_spaces(text,indeces_of_spaces)
+        return new_text
 
+    def replace_with_spaces(self, text, indices):
```
Comment (Eldar-Ka) [inline src/space_predictor/Iterative_space_predictor.py:34]: add docomentation please

**Review by Eldar-Ka — CHANGES_REQUESTED** [review]

File: src/model/Char_model.py [hunk 4]
```diff
@@ -68,7 +72,9 @@ def fill_word_preds(self, word: str, preds: list, p_lst: list) -> list:
 
         completions = []
         scores = []
-        for chars, score in zip(itertools.product(*char_lists), itertools.product(*p_lst)):
+        comb_chars=self.limit_product(char_lists,1000)#list(itertools.product(*char_lists))[:1000]
```
Comment (harelm4) [inline src/model/Char_model.py:75]: revert to old working version

File: src/model/ensemble_union.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        :param word_textpart: A TextPart of the word model
+        :param char_model_weight: the weight being used in the formula
+        :return: A merged TextPart
+        """
+        new_scores=self.Weighted_average(char_textpart.predictions, word_textpart.predictions, char_model_weight, 1-char_model_weight)
+        res=[Prediction(value,score) for value,score in new_scores.items()]
+        return TextPart('?', list(sorted(res, key=lambda x: x.score, reverse=True)))
+
+    def Weighted_average(self, list1, list2, weight1, weight2):
+
```
Comment (Eldar-Ka) [inline src/model/ensemble_union.py:43]: add docomentation please

**Review by Eldar-Ka — APPROVED** [review]

**Review by Eldar-Ka — APPROVED** [review]

**Pull request merged**